File: matlab/tests/ibllib/test_dsp.m
function tests = test_dsp
tests = functiontests(localfunctions);
end

function test_fscale(testCase)
%% dsp.fscale
res = [0;100;200;300;400;500;-400;-300;-200;-100];
assert(all(abs(dsp.fscale(10,0.001)-res)< 1e4))
assert(all(abs(dsp.fscale(11,0.001,'real')-res(1:6))< 1e4))

res = [0;90.9090909090909;181.818181818182;272.727272727273;363.636363636364;454.545454545455;-454.545454545455;-363.636363636364;-272.727272727273;-181.818181818182;-90.9090909090909];
assert(all(abs(dsp.fscale(11,0.001)-res)< 1e4))
assert(all(abs(dsp.fscale(11,0.001,'real')-res(1:6))< 1e4))
end

function test_freduce(testCase)
%% dsp.freduce
%test odd
res = [0;90.9090909090909;181.818181818182;272.727272727273;363.636363636364;454.545454545455;-454.545454545455;-363.636363636364;-272.727272727273;-181.818181818182;-90.9090909090909];
assert(all( dsp.freduce(res) == res(1:6)))
% test even
res = [0;100;200;300;400;500;-400;-300;-200;-100];
assert(all( dsp.freduce(res) == res(1:6)))
% test 2D
assert(all( flatten( dsp.freduce(repmat(res,1,2)) == repmat(res(1:6),1, 2))))
% test 3D
assert(all( flatten( dsp.freduce(repmat(res,1,2,3))  == repmat(res(1:6),1, 2,3))))
end

function test_expand(testCase)
%% dsp.fexpand
% test odd
res = rand(11,1);
X = dsp.freduce(fft(res));
R = real(ifft(dsp.fexpand(X,11)));
assert( all( abs(R - res) < 1e6))
R = real(ifft(dsp.fexpand(X))); % single arg
assert( all( abs(R - res) < 1e6))
% test even
res = rand(10,1);
X = dsp.freduce(fft(res));
X = real(ifft(dsp.fexpand(X,10)));
assert( all( abs(X - res) < 1e6))
% test 2D
res = rand(10,2);
X = dsp.freduce(fft(res));
X = real(ifft(dsp.fexpand(X,10)));
assert( all( abs(X(:) - res(:)) < 1e6))
% test 3D
res = rand(10,2,3);
X = dsp.freduce(fft(res));
X = real(ifft(dsp.fexpand(X,10)));
assert( all( abs(X(:) - res(:)) < 1e6))
end

%% Test ffilter
function test_ffilter(testCase)
ts = rand(500,1).*200+3400;
si = 0.001;
ts_ = dsp.ffilter.lp(ts, si, [150 200]);
% mean(ts) mean(ts_)
% dsp.Spectrum([ts ts_], 0.001)
testCase.assertTrue(abs(mean(ts)-mean(ts_)) < 1e-4)
ts = rand(500,2).*200+3400;
ts_ = dsp.ffilter.lp(ts, si, [150 200]);
testCase.assertTrue(all(abs(mean(ts)-mean(ts_)) < 1e-4))
end

%% Test Smooth
function test_smooth(testCase)
%%
Depth = [1:500]';
Vp = rand(500,1).*200+3400;
Vs = Vp*2/3;
Rho = ones(500,1)+1.5+rand(500,1).*0.1;
C = struct('Vp',Vp,'Vs',Vs);

% single serie input
[vp1] = dsp.smooth.lp(Vp, [0.1 0.2]);
assert(all(size(vp1)==size(Vp)))
[vp2] = dsp.smooth.mwa(Vp, 10);
assert(all(size(vp2)==size(Vp)))
[vp3] = dsp.smooth.exp_single(Vp, 0.01);
assert(all(size(vp3)==size(Vp)))

% array input
[vpvs1] = dsp.smooth.lp([Vp Vs], [0.1 0.2]);
assert(all(size(vpvs1)==size([Vp Vs])))
assert(all(vpvs1(:,1)==vp1))
[vpvs2] = dsp.smooth.mwa([Vp Vs], 10);
assert(all(size(vpvs2)==size([Vp Vs])))
assert(all(vpvs2(:,1)==vp2))
[vpvs3] = dsp.smooth.exp_single([Vp Vs], 0.1);
assert(all(size(vpvs3)==size([Vp Vs])))
assert(all(vpvs3(:,1)==vp3))

% structure input
[c1] = dsp.smooth.lp(C, [0.1 0.2]);
assert(all(flatten( cell2mat(struct2cell(c1)')==vpvs1 )))
[c2] = dsp.smooth.mwa(C, 10);
assert(all(flatten( cell2mat(struct2cell(c2)')==vpvs2 )))
[c3] = dsp.smooth.exp_single(C, 0.1);
assert(all(flatten( cell2mat(struct2cell(c3)')==vpvs3 )))
end

File: python/tests/ibllib/test_dsp.py
import unittest
import numpy as np
import numpy.matlib as mat

import ibllib.dsp.fourier as ft
from ibllib.dsp import WindowGenerator, rms, rises, falls, fronts


class TestFFT(unittest.TestCase):

    def test_imports(self):
        import ibllib.dsp as dsp
        self.assertTrue(len([dsp.lp,
                             dsp.fexpand,
                             dsp.hp,
                             dsp.fscale,
                             dsp.freduce,
                             dsp.rms]) == 6)

    def test_freduce(self):
        # test with 1D arrays
        fs = np.fft.fftfreq(5)
        self.assertTrue(np.all(ft.freduce(fs) == fs[:-2]))
        fs = np.fft.fftfreq(6)
        self.assertTrue(np.all(ft.freduce(fs) == fs[:-2]))

        # test 2D arrays along both dimensions
        fs = mat.repmat(ft.fscale(500, 0.001), 4, 1)
        self.assertTrue(ft.freduce(fs).shape == (4, 251))
        self.assertTrue(ft.freduce(np.transpose(fs), axis=0).shape == (251, 4))

    def test_fexpand(self):
        # test odd input
        res = np.random.rand(11)
        X = ft.freduce(np.fft.fft(res))
        R = np.real(np.fft.ifft(ft.fexpand(X, 11)))
        self.assertTrue(np.all((res - R) < 1e-6))
        # test even input
        res = np.random.rand(12)
        X = ft.freduce(np.fft.fft(res))
        R = np.real(np.fft.ifft(ft.fexpand(X, 12)))
        self.assertTrue(np.all((res - R) < 1e-6))
        # test with a 2 dimensional input along last dimension
        res = np.random.rand(2, 12)
        X = ft.freduce(np.fft.fft(res))
        R = np.real(np.fft.ifft(ft.fexpand(X, 12)))
        self.assertTrue(np.all((res - R) < 1e-6))
        # test with a 3 dimensional input along last dimension
        res = np.random.rand(3, 5, 12)
        X = ft.freduce(np.fft.fft(res))
        R = np.real(np.fft.ifft(ft.fexpand(X, 12)))
        self.assertTrue(np.all((res - R) < 1e-6))
        # test with 2 dimensional input along first dimension
        fs = np.transpose(mat.repmat(ft.fscale(500, 0.001, one_sided=True), 4, 1))
        self.assertTrue(ft.fexpand(fs, 500, axis=0).shape == (500, 4))

    def test_fscale(self):
        # test for an even number of samples
        res = [0, 100, 200, 300, 400, 500, -400, -300, -200, -100],
        self.assertTrue(np.all(np.abs(ft.fscale(10, 0.001) - res) < 1e-6))
        # test for an odd number of samples
        res = [0, 90.9090909090909, 181.818181818182, 272.727272727273, 363.636363636364,
               454.545454545455, -454.545454545455, -363.636363636364, -272.727272727273,
               -181.818181818182, -90.9090909090909],
        self.assertTrue(np.all(np.abs(ft.fscale(11, 0.001) - res) < 1e-6))

    def test_filter_lp_hp(self):
        # test 1D time serie: subtracting lp filter removes DC
        ts1 = np.random.rand(500)
        out1 = ft.lp(ts1, 1, [.1, .2])
        self.assertTrue(np.mean(ts1 - out1) < 0.001)
        # test 2D case along the last dimension
        ts = mat.repmat(ts1, 11, 1)
        out = ft.lp(ts, 1, [.1, .2])
        self.assertTrue(np.allclose(out, out1))
        # test 2D case along the first dimension
        ts = mat.repmat(ts1[:, np.newaxis], 1, 11)
        out = ft.lp(ts, 1, [.1, .2], axis=0)
        self.assertTrue(np.allclose(np.transpose(out), out1))
        # test 1D time serie: subtracting lp filter removes DC
        out2 = ft.hp(ts1, 1, [.1, .2])
        self.assertTrue(np.allclose(out1, ts1 - out2))


class TestWindowGenerator(unittest.TestCase):

    def test_window_simple(self):
        wg = WindowGenerator(ns=500, nswin=100, overlap=50)
        sl = list(wg.firstlast)
        self.assertTrue(wg.nwin == len(sl) == 9)
        self.assertTrue(np.all(np.array([s[0] for s in sl]) == np.arange(0, wg.nwin) * 50))
        self.assertTrue(np.all(np.array([s[1] for s in sl]) == np.arange(0, wg.nwin) * 50 + 100))

        wg = WindowGenerator(ns=500, nswin=100, overlap=10)
        sl = list(wg.firstlast)
        first = np.array([0, 90, 180, 270, 360, 450])
        last = np.array([100, 190, 280, 370, 460, 500])
        self.assertTrue(wg.nwin == len(sl) == 6)
        self.assertTrue(np.all(np.array([s[0] for s in sl]) == first))
        self.assertTrue(np.all(np.array([s[1] for s in sl]) == last))

    def test_nwindows_computation(self):
        for m in np.arange(0, 100):
            wg = WindowGenerator(ns=500 + m, nswin=87 + m, overlap=11 + m)
            sl = list(wg.firstlast)
            self.assertTrue(wg.nwin == len(sl))

    def test_firstlast_slices(self):
        # test also the indexing versus direct slicing
        my_sig = np.random.rand(500,)
        wg = WindowGenerator(ns=500, nswin=100, overlap=50)
        # 1) get the window by
        my_rms = np.zeros((wg.nwin,))
        for first, last in wg.firstlast:
            my_rms[wg.iw] = rms(my_sig[first:last])
        # test with slice_array method
        my_rms_ = np.zeros((wg.nwin,))
        for wsig in wg.slice_array(my_sig):
            my_rms_[wg.iw] = rms(wsig)
        self.assertTrue(np.all(my_rms_ == my_rms))
        # test with the slice output
        my_rms_ = np.zeros((wg.nwin,))
        for sl in wg.slice:
            my_rms_[wg.iw] = rms(my_sig[sl])
        self.assertTrue(np.all(my_rms_ == my_rms))

    def test_tscale(self):
        wg = WindowGenerator(ns=500, nswin=100, overlap=50)
        ts = wg.tscale(fs=1000)
        self.assertTrue(ts[0] == (100 - 1) / 2 / 1000)
        self.assertTrue((np.allclose(np.diff(ts), 0.05)))

    def test_rises_falls(self):
        # test 1D case with a long pulse and a dirac
        a = np.zeros(500,)
        a[80:120] = 1
        a[200] = 1
        # rising fronts
        self.assertTrue(all(rises(a) == np.array([80, 200])))
        # falling fronts
        self.assertTrue(all(falls(a) == np.array([120, 201])))
        # both
        ind, val = fronts(a)
        self.assertTrue(all(ind == np.array([80, 120, 200, 201])))
        self.assertTrue(all(val == np.array([1, -1, 1, -1])))

        # test a 2D case with 2 long pulses and a dirac
        a = np.zeros((2, 500))
        a[0, 80:120] = 1
        a[0, 200] = 1
        a[1, 280:320] = 1
        a[1, 400] = 1
        # rising fronts
        self.assertTrue(np.all(rises(a) == np.array([[0, 0, 1, 1], [80, 200, 280, 400]])))
        # falling fronts
        self.assertTrue(np.all(falls(a) == np.array([[0, 0, 1, 1], [120, 201, 320, 401]])))
        # both
        ind, val = fronts(a)
        self.assertTrue(all(ind[0] == np.array([0, 0, 0, 0, 1, 1, 1, 1])))
        self.assertTrue(all(ind[1] == np.array([80, 120, 200, 201, 280, 320, 400, 401])))
        self.assertTrue(all(val == np.array([1, -1, 1, -1, 1, -1, 1, -1])))


if __name__ == "__main__":
    unittest.main(exit=False)
